### ISSUE TITLE
Hosts are getting rebooted, although there was nothing to do.

### ISSUE THREAD

**Issue by Tronde** [Tronde/ansible-role-rhel-patchmanagement#12]
To determine whether some host needs to be rebooted or not the output of the yum/dnf command is checked as follows:

```
  - name: Reboot Host if packages were updated
    shell: sleep 2 && shutdown -r now "Ansible updates triggered"
    async: 1
    poll: 0
    ignore_errors: true
    when: ("Complete!" in yum_output.stdout_lines[-1]) or
          ("Komplett!" in yum_output.stdout_lines[-1])
```
But the strings "Complete!" or "Komplett!" occur in RHEL 8 even when there is nothing to do, as shown in the following debug output:

```
TASK [rhel-patchmanagement : debug] ********************************************
ok: [rhel8.example.com] => {
    "yum_output": {
        "changed": true,
[...]
        "stderr": "",
        "stderr_lines": [],
        "stdout": "Updating Subscription Management repositories.\nLast metadata expiration check: 3:19:20 ago on Wed 29 Dec 2021 01:01:03 AM CET.\nDependencies resolved.\nNothing to do.\nComplete!",
        "stdout_lines": [
            "Updating Subscription Management repositories.",
            "Last metadata expiration check: 3:19:20 ago on Wed 29 Dec 2021 01:01:03 AM CET.",
            "Dependencies resolved.",
            "Nothing to do.",
            "Complete!"
        ],
    }
}
```

An already patched RHEL 7 host shows:

```
ok: [rhel7.example.com] => {
    "yum_output": {
[...]
        "stderr": "",
        "stderr_lines": [],
        "stdout": "Loaded plugins: product-id, search-disabled-repos, subscription-manager\nNo Packages marked for minimal Update",
        "stdout_lines": [
            "Loaded plugins: product-id, search-disabled-repos, subscription-manager",
            "No Packages marked for minimal Update"
        ],
    }
}
```

## Ideas

 * Check if there are similar strings other than "Complete!" or "Komplett!" for RHEL 7 and RHEL 8 that could be used to check whether a reboot is needed or not.
 * Alternative: Write different reboot conditions for RHEL 7 and RHEL 8.
 * Check RHEL 9 as well.

**Comment by Tronde** [comment]
# Information gathering from latest patch run
I checked the latest log and extracted `stdout_lines[]` for different situations.

## RHEL 7 had no packages marked for minimal update
~~~
        "stdout_lines": [
            "Loaded plugins: product-id, search-disabled-repos, subscription-manager",
            "No Packages marked for minimal Update"
        ],
~~~

## RHEL 7 had packages marked for minimal update
~~~
stdout_lines[-1] == "Complete!"
~~~

## RHEL 8 had packages marked for minimal update
~~~
        "stdout_lines": [
            "Updating Subscription Management repositories.",
            "Last metadata expiration check: 0:14:21 ago on Tue 11 Jan 2022 04:07:03 AM CET.",
            "Dependencies resolved.",
            ...
            "Complete!"
        ],
~~~

## RHEL 8 had no packages marked for minimal update
~~~
        "stdout_lines": [
            "Updating Subscription Management repositories.",
            "Last metadata expiration check: 2:02:13 ago on Tue 11 Jan 2022 02:21:43 AM CET.",
            "Dependencies resolved.",
            "Nothing to do.",
            "Complete!"
        ],
~~~

## Information for RHEL 9 Beta
~~~
rhel-9-beta-1 ~]$ sudo dnf update-minimal
Updating Subscription Management repositories.
Last metadata expiration check: 0:34:27 ago on Di 11 Jan 2022 08:23:26 CET.
No security updates needed, but 154 updates available
Dependencies resolved.
Nothing to do.
Complete!
~~~

# Possible solution
The only possible solution seems to be to use different conditons for the different major relases as there is no common string otherwise to match and decide whether a reboot is needed or not.

**Comment by Klaas-** [comment]
you can use needs-restarting, but it uses shell/command

```
- name: check if reboot is needed
  shell: "{{ needs_restarting_command[ansible_distribution_major_version | int] }}"
  failed_when: False
  register: needs_restarting
  changed_when: False
```

definition of the var
```
needs_restarting_command:
  7: /usr/bin/needs-restarting -r
  8: /usr/bin/dnf needs-restarting -r
```

usage in the reboot task:
`when: needs_restarting.rc is defined and needs_restarting.rc != 0`

on rhel7 deps are `yum-utils`, on rhel8 `python3-dnf-plugins-core`

**Comment by Tronde** [comment]
Hi @Klaas- and thank you very much for you comment. I'll consider your suggestion for the next test run. When it works for me, I'm going to add it to the tasks of the role.

**Comment by Tronde** [comment]
Closed with #13.
